typed/clj.checker/src/typed/cljc/checker/type_ctors.clj

```diff
@@ -77,7 +77,10 @@
   ([s hint]
    {:pre [(symbol? s)]
     :post [(symbol? %)]}
-   (with-meta (gensym (str s hint)) {:original-name s})))
+   (-> (.concat (str s) (str hint))
+       (.concat (str (clojure.lang.RT/nextID)))
+       clojure.lang.Symbol/intern
+       (with-meta {:original-name s}))))
```

Comment:
Micro-optimization ideas:

- I think the 1-arity is the hottest path (via the `*-free-names*` functions), might be worth duplicating the logic.
  - I'm guessing `.concat` is linear, so we'd just need to `.concat` once there. Might be worth considering using our own StringBuilder (or something lower level)?
- `s` is a simple-symbol so maybe `(.-name ^Symbol s)` instead of `(str s)`.

Reply:
Once concat call is identical to creating a StringBuilder for it; and at this point it's not worth complicating this further. I've duplicated the logic between arities like you've suggested.

Single-arity `str` is fine, `(.-name ^Symbol s)` wouldn't improve it much.

Reply:
I made a couple of tweaks since the 2-arg intern doesn't walk the name looking for `/` and Integer/toString seems to be the fundamental operation for integers without any virtual calls to compile away.

Agreed `.-name` is overkill, I like `name` for style here though.